dds/DCPS/RTPS/Sedp.cpp

```diff
@@ -936,18 +950,61 @@ void Sedp::associate_secure_readers_to_writers(const Security::SPDPdiscoveredPar
 }
 
 void
-Sedp::send_builtin_crypto_tokens(const DCPS::RepoId& dstParticipant, const DCPS::EntityId_t& dstEntity, const DCPS::RepoId& src)
+Sedp::create_and_send_datareader_crypto_tokens(
+  const DDS::Security::DatareaderCryptoHandle& drch, const DCPS::RepoId& local_reader,
+  const DDS::Security::DatawriterCryptoHandle& dwch, const DCPS::RepoId& remote_writer)
+{
+  DCPS::RepoId remote_participant;
+  std::memcpy(remote_participant.guidPrefix, remote_writer.guidPrefix, sizeof(GuidPrefix_t));
+  remote_participant.entityId = ENTITYID_PARTICIPANT;
+
+  RepeatOnceWriter::RemoteWriter info;
+  info.local_reader = local_reader;
+  info.remote_writer = remote_writer;
+  DDS::Security::DatareaderCryptoTokenSeq& drcts = info.reader_tokens;
+  create_datareader_crypto_tokens(drch, dwch, drcts);
+
+  send_datareader_crypto_tokens(local_reader, remote_writer, drcts);
+  ACE_Guard<ACE_Thread_Mutex> g(participant_volatile_message_secure_writer_.lock_);
+  participant_volatile_message_secure_writer_.
+    remote_writers_[remote_participant].push_back(info);
+}
+
+void
+Sedp::create_and_send_datawriter_crypto_tokens(
+  const DDS::Security::DatawriterCryptoHandle& dwch, const DCPS::RepoId& local_writer,
+  const DDS::Security::DatareaderCryptoHandle& drch, const DCPS::RepoId& remote_reader)
+{
+  DCPS::RepoId remote_participant;
+  std::memcpy(remote_participant.guidPrefix, remote_reader.guidPrefix, sizeof(GuidPrefix_t));
+  remote_participant.entityId = ENTITYID_PARTICIPANT;
+
+  RepeatOnceWriter::RemoteReader info;
+  info.local_writer = local_writer;
+  info.remote_reader = remote_reader;
+  DDS::Security::DatawriterCryptoTokenSeq& dwcts = info.writer_tokens;
+  create_datawriter_crypto_tokens(dwch, drch, dwcts);
+
+  send_datawriter_crypto_tokens(local_writer, remote_reader, dwcts);
+  ACE_Guard<ACE_Thread_Mutex> g(participant_volatile_message_secure_writer_.lock_);
+  participant_volatile_message_secure_writer_.
+    remote_readers_[remote_participant].push_back(info);
```

Comment:
So if the ACK never shows up, do these entries just stick around forever? I'm guessing we'll need to clear these out on participant unregister / dispose or we might wind up with collisions on reconnect. Are we clearing these out somewhere?

Reply:
> So if the ACK never shows up, do these entries just stick around forever?

Yes

> I'm guessing we'll need to clear these out on participant unregister / dispose or we might wind up with collisions on reconnect. Are we clearing these out somewhere?

No. We don't do any clean up for keys on participant dispose right now. When we do, clearing any pending key resends would probably be a good idea.